src/tablet/data_receiver.cc

```diff
@@ -0,0 +1,122 @@
+        data_blocks_.emplace_back(new storage::DataBlock(info.ref_cnt(), buf, info.length(), true));
+    }
+    if (iter.bytes_left() != 0) {
+        LOG(ERROR) << tid_ << "-" << pid_ << " data and info mismatch, revert this part";
+        // Not only ptr, DataBlock needs delete.
+        for (auto i = last_block_size; i < data_blocks_.size(); ++i) {
+            // regardless of dim_cnt_down
+            delete data_blocks_[i];
+        }
+        // range erase
+        data_blocks_.resize(last_block_size);
+        return false;
+    }
+
+    DLOG(INFO) << "inserted into table(" << tid_ << "-" << pid_ << ") " << request->block_info_size()
+               << " rows. Looking forward to part " << next_part_id_ << " or IndexRegion.";
+    return true;
+}
+
+bool DataReceiver::PartValidation(int part_id) {
+    if (part_id != next_part_id_) {
+        LOG(WARNING) << tid_ << "-" << pid_ << " data receiver needs part " << next_part_id_ << ", but get part "
+                     << part_id;
+        return false;
+    }
+    next_part_id_++;
+    return true;
+}
+
+bool DataReceiver::BulkLoad(std::shared_ptr<storage::MemTable> table, const ::openmldb::api::BulkLoadRequest* request) {
+    std::unique_lock<std::mutex> ul(mu_);
+    DLOG_ASSERT(tid_ == table->GetId() && pid_ == table->GetPid());
+
+    if (!request->has_part_id() || !PartValidation(request->part_id())) {
+        LOG(WARNING) << tid_ << "-" << pid_ << " data receiver received invalid part id, expect " << next_part_id_
+                     << ", actual " << (request->has_part_id() ? "no id" : std::to_string(request->part_id()));
+        return false;
+    }
+
+    if (!table->BulkLoad(data_blocks_, request->index_region())) {
+        LOG(ERROR) << "bulk load to mem table(" << tid_ << "-" << pid_ << ") failed.";
+    }
+
+    DLOG(INFO) << "bulk load to mem table(" << tid_ << "-" << pid_ << ") " << data_blocks_.size() << " rows.";
+    return true;
+}
+
+bool DataReceiver::WriteBinlogToReplicator(const std::shared_ptr<replica::LogReplicator>& replicator,
+                                           const ::openmldb::api::BulkLoadRequest* request) {
+    std::unique_lock<std::mutex> ul(mu_);
+    // Do not do PartValidation
+    // TODO(hw): maybe binlog should hava the part id too?
+    if (request->part_id() != next_part_id_ - 1) {
+        LOG(WARNING)
+            << "WriteBinlogToReplicator follows AppendData, the same request, the same part id, but cur part id"
+            << next_part_id_ - 1 << ", request part id " << request->part_id();
+        return false;
+    }
+    for (int i = 0; i < request->binlog_info_size(); ++i) {
+        const auto& info = request->binlog_info(i);
+        ::openmldb::api::LogEntry entry;
+        auto* block = info.block_id() < data_blocks_.size() ? data_blocks_[info.block_id()] : nullptr;
+        if (block == nullptr) {
+            LOG(ERROR) << "binlog wants " << info.block_id() << ", but cached block size = " << data_blocks_.size();
```

Comment:
WARNING

Reply:
It's an error, not a warning which can be ignored